From 2bf627110474647a9cb72f463fa1ca178ee7bab7 Mon Sep 17 00:00:00 2001
From: redducc <sushmit718@gmail.com>
Date: Fri, 25 Sep 2026 11:01:54 +0000
Subject: [PATCH] SystemUI: Add a HarmonyOS Control Centre QS style [1/2]

Media beside Wi-Fi and Bluetooth, round toggles that expand from a handle, and cast and connected device cards.
---
 .../qs/ui/composable/QuickSettingsContent.kt  |  92 ++++
 .../shade/ui/composable/ShadeScene.kt         |  20 +-
 .../SystemUI/res/values/custom_strings.xml    |  11 +-
 .../composefragment/HarmonyQuickSettings.kt   | 434 ++++++++++++++++++
 .../systemui/qs/panels/ui/compose/EditMode.kt |  21 +
 .../qs/panels/ui/compose/HarmonyEditMode.kt   | 328 +++++++++++++
 .../systemui/qs/shared/style/QsPanelStyle.kt  |   4 +-
 7 files changed, 901 insertions(+), 9 deletions(-)
 create mode 100644 packages/SystemUI/src/com/android/systemui/qs/composefragment/HarmonyQuickSettings.kt
 create mode 100644 packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/HarmonyEditMode.kt

diff --git a/packages/SystemUI/compose/features/src/com/android/systemui/qs/ui/composable/QuickSettingsContent.kt b/packages/SystemUI/compose/features/src/com/android/systemui/qs/ui/composable/QuickSettingsContent.kt
index 07a9d1ede447..9ca47642579d 100644
--- a/packages/SystemUI/compose/features/src/com/android/systemui/qs/ui/composable/QuickSettingsContent.kt
+++ b/packages/SystemUI/compose/features/src/com/android/systemui/qs/ui/composable/QuickSettingsContent.kt
@@ -79,6 +79,13 @@ import com.android.systemui.qs.composefragment.POSITION_BELOW_GRID
 import com.android.systemui.qs.composefragment.POSITION_HEADER
 import com.android.systemui.qs.composefragment.SETTING_QS_FOLDER_POSITION
 import com.android.systemui.qs.composefragment.ConnectivityFolderExpansion
+import com.android.systemui.qs.composefragment.harmonyCardSpecs
+import com.android.systemui.qs.composefragment.HarmonyTopRow
+import com.android.systemui.qs.composefragment.HarmonyTogglesCard
+import com.android.systemui.qs.composefragment.HarmonyTitle
+import com.android.systemui.qs.composefragment.HarmonyGap
+import com.android.systemui.qs.composefragment.HarmonyConnectedDevices
+import com.android.systemui.qs.composefragment.HarmonyCastCard
 import com.android.systemui.qs.composefragment.SETTING_QS_FOLDER_SPAN
 import com.android.systemui.qs.composefragment.SETTING_QS_MEDIA_POSITION
 import com.android.systemui.qs.panels.ui.viewmodel.TileViewModel
@@ -113,6 +120,7 @@ import com.android.systemui.qs.shared.style.QsPanelStyle
 import com.android.systemui.qs.shared.ui.QuickSettings.Elements
 import com.android.systemui.qs.ui.viewmodel.QuickSettingsContainerViewModel
 import com.android.systemui.res.R
+import com.android.systemui.scene.shared.model.Overlays
 import kotlinx.coroutines.flow.filterNotNull
 
 internal class QsHostTransition(val idle: Boolean, val leaving: Boolean)
@@ -145,6 +153,7 @@ fun ContentScope.QuickSettingsContent(
                 PenguinQuickSettingsContent(viewModel, mediaInRow, modifier, mediaSquishiness)
             QsPanelStyle.MyUi ->
                 MyUiQuickSettingsContent(viewModel, modifier, mediaSquishiness)
+            QsPanelStyle.Harmony -> HarmonyQuickSettingsContent(viewModel, modifier)
         }
     }
 }
@@ -209,6 +218,89 @@ private fun ContentScope.MyUiQuickSettingsContent(
     }
 }
 
+@Composable
+private fun ContentScope.HarmonyQuickSettingsContent(
+    viewModel: QuickSettingsContainerViewModel,
+    modifier: Modifier = Modifier,
+) {
+    val allTiles = viewModel.tileGridViewModel.tileViewModels
+    val cardSpecs = remember(allTiles) { harmonyCardSpecs(allTiles) }
+    val toggles = remember(allTiles, cardSpecs) { allTiles.filterNot { it.spec.spec in cardSpecs } }
+    var interactable by remember { mutableStateOf(false) }
+    LaunchedEffect(Unit) {
+        snapshotFlow { Elements.QuickSettingsContent.currentAlpha() }
+            .filterNotNull()
+            .collect { interactable = it >= .5f }
+    }
+
+    Column(
+        verticalArrangement = spacedBy(HarmonyGap),
+        modifier =
+            modifier
+                .element(Elements.QuickSettingsContent)
+                .padding(horizontal = dimensionResource(id = R.dimen.qs_horizontal_margin))
+                .sysuiResTag("quick_settings_panel"),
+    ) {
+        HarmonyTitle()
+        HarmonyHeader(viewModel)
+        Box(Modifier.element(Elements.QuickSettingsTiles)) {
+            GridAnchor()
+            HarmonyTogglesCard(
+                tiles = toggles,
+                brightness = {
+                    Element(key = Elements.BrightnessSlider, modifier = Modifier) {
+                        Box(Modifier.thenIf(!interactable) { Modifier.gesturesDisabled() }) {
+                            BrightnessLayout(
+                                enable = interactable,
+                                horizontal = true,
+                                sliderHeight = LyingSliderHeight,
+                            )
+                        }
+                    }
+                },
+            )
+        }
+        HarmonyCastCard(allTiles)
+        HarmonyConnectedDevices(allTiles)
+        if (contentKey != Overlays.QuickSettingsShade) {
+            val editButtonViewModel =
+                rememberViewModel(traceName = "HarmonyQuickSettings-editButton") {
+                    viewModel.editModeButtonViewModelFactory.create()
+                }
+            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
+                EditModeButton(viewModel = editButtonViewModel, isVisible = interactable)
+            }
+        }
+    }
+}
+
+@Composable
+fun ContentScope.HarmonyHeader(
+    viewModel: QuickSettingsContainerViewModel,
+    modifier: Modifier = Modifier,
+) {
+    val showMedia =
+        viewModel.showMedia && viewModel.hasMediaCards && isAlwaysComposedContentVisible()
+    HarmonyTopRow(
+        tiles = viewModel.tileGridViewModel.tileViewModels,
+        media =
+            if (showMedia) {
+                {
+                    Element(key = Media.Elements.MediaCarousel, modifier = Modifier) {
+                        PenguinMediaCard(
+                            viewModelFactory = viewModel.mediaViewModelFactory,
+                            behavior = QuickSettingsContainerViewModel.mediaUiBehavior,
+                            square = true,
+                        )
+                    }
+                }
+            } else {
+                null
+            },
+        modifier = modifier.element(Elements.ConnectivityFolder),
+    )
+}
+
 @Composable
 fun ContentScope.MyUiHeaderRow(
     tiles: List<TileViewModel>,
diff --git a/packages/SystemUI/compose/features/src/com/android/systemui/shade/ui/composable/ShadeScene.kt b/packages/SystemUI/compose/features/src/com/android/systemui/shade/ui/composable/ShadeScene.kt
index c80541c84ea8..607215a84754 100644
--- a/packages/SystemUI/compose/features/src/com/android/systemui/shade/ui/composable/ShadeScene.kt
+++ b/packages/SystemUI/compose/features/src/com/android/systemui/shade/ui/composable/ShadeScene.kt
@@ -19,6 +19,7 @@ package com.android.systemui.shade.ui.composable
 import com.android.systemui.qs.composefragment.BrightnessLayout
 import com.android.systemui.qs.ui.composable.qsHostTransition
 import com.android.systemui.qs.ui.composable.LocalQsHostTransition
+import com.android.systemui.qs.ui.composable.HarmonyHeader
 import com.android.systemui.qs.composefragment.ConnectivityFolder
 import com.android.systemui.qs.composefragment.MyUiGridGap
 import com.android.systemui.qs.composefragment.MyUiMediaCard
@@ -406,7 +407,8 @@ private fun ContentScope.SingleShade(
             },
             mediaAndQqsHeader = {
                 val isDefaultStyle = viewModel.panelStyle == QsPanelStyle.Default
-                val isMyUiStyle = viewModel.panelStyle == QsPanelStyle.MyUi
+                val isHarmonyStyle = viewModel.panelStyle == QsPanelStyle.Harmony
+                val isMyUiStyle = viewModel.panelStyle == QsPanelStyle.MyUi || isHarmonyStyle
                 val qqsShowsMedia =
                     !isDefaultStyle &&
                         viewModel.isQsEnabled &&
@@ -694,13 +696,17 @@ private fun ContentScope.SingleShade(
                         }
                     },
                     myUiHeader = {
-                        MyUiHeaderRow(
-                            tiles =
-                                viewModel.qsContainerViewModel.tileGridViewModel.tileViewModels,
-                            interactable = true,
-                        )
+                        if (isHarmonyStyle) {
+                            HarmonyHeader(viewModel.qsContainerViewModel)
+                        } else {
+                            MyUiHeaderRow(
+                                tiles =
+                                    viewModel.qsContainerViewModel.tileGridViewModel.tileViewModels,
+                                interactable = true,
+                            )
+                        }
                     },
-                    showMedia = qqsShowsMedia,
+                    showMedia = qqsShowsMedia && !isHarmonyStyle,
                 )
                 }
             },
diff --git a/packages/SystemUI/res/values/custom_strings.xml b/packages/SystemUI/res/values/custom_strings.xml
index c276b088564b..7adc6b94641e 100644
--- a/packages/SystemUI/res/values/custom_strings.xml
+++ b/packages/SystemUI/res/values/custom_strings.xml
@@ -230,5 +230,14 @@
     <string name="weather_condition_snow">Snowy</string>
     <string name="weather_condition_wind">Windy</string>
     <string name="weather_condition_mist">Misty</string>
-</resources>
 
+    <!-- HarmonyOS Control Centre -->
+    <string name="harmony_control_centre">Control centre</string>
+    <string name="harmony_not_playing">Not playing</string>
+    <string name="harmony_devices">Devices</string>
+    <string name="harmony_cast_hint">Tap to cast your screen to a nearby device</string>
+    <string name="harmony_connected">Connected</string>
+    <string name="harmony_edit_title">Edit control centre</string>
+    <string name="harmony_edit_subtitle">Long press and drag to rearrange</string>
+    <string name="harmony_edit_more">More toggles</string>
+</resources>
diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/HarmonyQuickSettings.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/HarmonyQuickSettings.kt
new file mode 100644
index 000000000000..7fff3352704e
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/HarmonyQuickSettings.kt
@@ -0,0 +1,434 @@
+/*
+ * Copyright (C) 2026 The PenguinOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.systemui.qs.composefragment
+
+import android.service.quicksettings.Tile
+import androidx.compose.animation.animateColorAsState
+import androidx.compose.animation.animateContentSize
+import androidx.compose.foundation.background
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.combinedClickable
+import androidx.compose.foundation.gestures.detectVerticalDragGestures
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Arrangement.spacedBy
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.layout.width
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.filled.ArrowDropDown
+import androidx.compose.material.icons.filled.MusicNote
+import androidx.compose.material3.Icon as MaterialIcon
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.text.style.TextOverflow
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.compose.LifecycleStartEffect
+import com.android.systemui.common.shared.model.Icon
+import com.android.systemui.qs.panels.ui.compose.TileListener
+import com.android.systemui.qs.panels.ui.compose.infinitegrid.SmallTileContent
+import com.android.systemui.qs.panels.ui.viewmodel.TileViewModel
+import com.android.systemui.res.R
+
+internal val HarmonyGap = 12.dp
+private val HarmonyCorner = 24.dp
+private val HarmonyAccent = Color(0xFF0A59F7)
+private val HarmonyToggleSize = 52.dp
+private const val HarmonyToggleColumns = 5
+
+private val WifiSpecs = listOf("wifi", "internet")
+private const val BluetoothSpec = "bt"
+private const val CastSpec = "cast"
+
+fun harmonyCardSpecs(tiles: List<TileViewModel>): Set<String> {
+    val present = tiles.map { it.spec.spec }.toSet()
+    return buildSet {
+        WifiSpecs.firstOrNull { it in present }?.let(::add)
+        if (BluetoothSpec in present) add(BluetoothSpec)
+        if (CastSpec in present) add(CastSpec)
+    }
+}
+
+@Composable
+private fun Modifier.harmonyCard(): Modifier =
+    clip(RoundedCornerShape(HarmonyCorner)).background(glassSurface())
+
+@Composable
+private fun ListenTo(tiles: List<TileViewModel>) {
+    var listening by remember { mutableStateOf(false) }
+    LifecycleStartEffect(Unit) {
+        listening = true
+        onStopOrDispose { listening = false }
+    }
+    TileListener(tiles) { listening }
+}
+
+@Composable
+fun HarmonyTitle(modifier: Modifier = Modifier) {
+    Text(
+        text = stringResource(R.string.harmony_control_centre),
+        color = MaterialTheme.colorScheme.onSurface,
+        style = MaterialTheme.typography.headlineSmall,
+        modifier = modifier.padding(horizontal = 4.dp, vertical = 4.dp),
+    )
+}
+
+@Composable
+fun HarmonyTopRow(
+    tiles: List<TileViewModel>,
+    media: (@Composable () -> Unit)?,
+    modifier: Modifier = Modifier,
+) {
+    val bySpec = remember(tiles) { tiles.associateBy { it.spec.spec } }
+    val wifi = WifiSpecs.firstNotNullOfOrNull { bySpec[it] }
+    val bluetooth = bySpec[BluetoothSpec]
+    ListenTo(listOfNotNull(wifi, bluetooth))
+    val height = qsModuleHeight(2)
+    Row(
+        modifier = modifier.fillMaxWidth().height(height),
+        horizontalArrangement = spacedBy(HarmonyGap),
+    ) {
+        Box(Modifier.weight(1f).fillMaxSize()) { media?.invoke() ?: HarmonyIdleMedia() }
+        Column(Modifier.weight(1f).fillMaxSize(), verticalArrangement = spacedBy(HarmonyGap)) {
+            wifi?.let { HarmonyNetworkCard(it, Modifier.weight(1f)) }
+            bluetooth?.let { HarmonyNetworkCard(it, Modifier.weight(1f), dropDown = true) }
+        }
+    }
+}
+
+@Composable
+private fun HarmonyIdleMedia() {
+    Column(
+        modifier = Modifier.fillMaxSize().harmonyCard().padding(16.dp),
+        verticalArrangement = Arrangement.SpaceBetween,
+    ) {
+        Box(
+            modifier =
+                Modifier.size(44.dp)
+                    .clip(RoundedCornerShape(12.dp))
+                    .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.08f)),
+            contentAlignment = Alignment.Center,
+        ) {
+            MaterialIcon(
+                imageVector = Icons.Filled.MusicNote,
+                contentDescription = null,
+                tint = MaterialTheme.colorScheme.onSurfaceVariant,
+            )
+        }
+        Text(
+            text = stringResource(R.string.harmony_not_playing),
+            color = MaterialTheme.colorScheme.onSurface,
+            style = MaterialTheme.typography.titleSmall,
+        )
+    }
+}
+
+@Composable
+private fun HarmonyNetworkCard(
+    tile: TileViewModel,
+    modifier: Modifier = Modifier,
+    dropDown: Boolean = false,
+) {
+    val (uiState, icon) = rememberTileState(tile)
+    val active = uiState.visualState == Tile.STATE_ACTIVE
+    Row(
+        modifier =
+            modifier
+                .fillMaxWidth()
+                .harmonyCard()
+                .combinedClickable(
+                    onClick = { tile.primaryAction(uiState) },
+                    onLongClick = { tile.settingsClick(null) },
+                )
+                .padding(start = 12.dp, end = 8.dp),
+        horizontalArrangement = spacedBy(10.dp),
+        verticalAlignment = Alignment.CenterVertically,
+    ) {
+        HarmonyCircle(icon = icon, active = active, size = 40.dp)
+        Column(Modifier.weight(1f)) {
+            Text(
+                text = uiState.label,
+                color = MaterialTheme.colorScheme.onSurface,
+                style = MaterialTheme.typography.titleSmall,
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis,
+            )
+            if (uiState.secondaryLabel.isNotBlank()) {
+                Text(
+                    text = uiState.secondaryLabel,
+                    color = MaterialTheme.colorScheme.onSurfaceVariant,
+                    style = MaterialTheme.typography.bodySmall,
+                    maxLines = 1,
+                    overflow = TextOverflow.Ellipsis,
+                )
+            }
+        }
+        if (dropDown) {
+            Box(
+                modifier =
+                    Modifier.size(32.dp).clip(CircleShape).clickable { tile.mainClick(null) },
+                contentAlignment = Alignment.Center,
+            ) {
+                MaterialIcon(
+                    imageVector = Icons.Filled.ArrowDropDown,
+                    contentDescription = null,
+                    tint = MaterialTheme.colorScheme.onSurface,
+                )
+            }
+        }
+    }
+}
+
+@Composable
+private fun HarmonyCircle(icon: Icon, active: Boolean, size: Dp) {
+    val background by
+        animateColorAsState(
+            if (active) HarmonyAccent else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.1f),
+            label = "HarmonyCircle",
+        )
+    Box(
+        modifier = Modifier.size(size).clip(CircleShape).background(background),
+        contentAlignment = Alignment.Center,
+    ) {
+        SmallTileContent(
+            iconProvider = { icon },
+            color = if (active) Color.White else MaterialTheme.colorScheme.onSurface,
+            size = { size * 0.46f },
+        )
+    }
+}
+
+@Composable
+fun HarmonyTogglesCard(
+    tiles: List<TileViewModel>,
+    brightness: @Composable () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    ListenTo(tiles)
+    var expanded by remember { mutableStateOf(false) }
+    LifecycleStartEffect(Unit) { onStopOrDispose { expanded = false } }
+    val canExpand = tiles.size > HarmonyToggleColumns
+    val shown = if (expanded) tiles else tiles.take(HarmonyToggleColumns)
+
+    Column(
+        modifier =
+            modifier
+                .fillMaxWidth()
+                .harmonyCard()
+                .animateContentSize()
+                .padding(start = 14.dp, end = 14.dp, top = 16.dp),
+        verticalArrangement = spacedBy(14.dp),
+    ) {
+        shown.chunked(HarmonyToggleColumns).forEach { row ->
+            Row(modifier = Modifier.fillMaxWidth()) {
+                row.forEach { HarmonyToggle(it, labelled = expanded, Modifier.weight(1f)) }
+                repeat(HarmonyToggleColumns - row.size) { Box(Modifier.weight(1f)) }
+            }
+        }
+        brightness()
+        Box(
+            modifier =
+                Modifier.fillMaxWidth()
+                    .height(24.dp)
+                    .then(
+                        if (canExpand) {
+                            Modifier.clickable { expanded = !expanded }
+                                .pointerInput(Unit) {
+                                    detectVerticalDragGestures { change, dragAmount ->
+                                        if (dragAmount > 8f) expanded = true
+                                        else if (dragAmount < -8f) expanded = false
+                                        change.consume()
+                                    }
+                                }
+                        } else {
+                            Modifier
+                        }
+                    ),
+            contentAlignment = Alignment.Center,
+        ) {
+            if (canExpand) {
+                Box(
+                    Modifier.width(28.dp)
+                        .height(4.dp)
+                        .clip(CircleShape)
+                        .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.25f))
+                )
+            }
+        }
+    }
+}
+
+@Composable
+private fun HarmonyToggle(tile: TileViewModel, labelled: Boolean, modifier: Modifier = Modifier) {
+    val (uiState, icon) = rememberTileState(tile)
+    Column(
+        modifier =
+            modifier.combinedClickable(
+                onClick = { tile.primaryAction(uiState) },
+                onLongClick = { tile.settingsClick(null) },
+            ),
+        horizontalAlignment = Alignment.CenterHorizontally,
+        verticalArrangement = spacedBy(6.dp),
+    ) {
+        HarmonyCircle(icon, uiState.visualState == Tile.STATE_ACTIVE, HarmonyToggleSize)
+        if (labelled) {
+            Text(
+                text = uiState.label,
+                color = MaterialTheme.colorScheme.onSurface,
+                style = MaterialTheme.typography.labelSmall,
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis,
+                textAlign = TextAlign.Center,
+            )
+        }
+    }
+}
+
+@Composable
+fun HarmonyCastCard(tiles: List<TileViewModel>, modifier: Modifier = Modifier) {
+    val cast = remember(tiles) { tiles.firstOrNull { it.spec.spec == CastSpec } } ?: return
+    ListenTo(listOf(cast))
+    val (uiState, icon) = rememberTileState(cast)
+    val active = uiState.visualState == Tile.STATE_ACTIVE
+    Column(
+        modifier =
+            modifier
+                .fillMaxWidth()
+                .harmonyCard()
+                .clickable { cast.mainClick(null) }
+                .padding(16.dp),
+        verticalArrangement = spacedBy(12.dp),
+        horizontalAlignment = Alignment.CenterHorizontally,
+    ) {
+        Text(
+            text = stringResource(R.string.harmony_devices),
+            color = MaterialTheme.colorScheme.onSurface,
+            style = MaterialTheme.typography.titleSmall,
+            modifier = Modifier.fillMaxWidth(),
+        )
+        Box(
+            modifier =
+                Modifier.size(64.dp)
+                    .clip(CircleShape)
+                    .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.06f)),
+            contentAlignment = Alignment.Center,
+        ) {
+            HarmonyCircle(icon = icon, active = active, size = 44.dp)
+        }
+        Text(
+            text =
+                if (active && uiState.secondaryLabel.isNotBlank()) uiState.secondaryLabel
+                else stringResource(R.string.harmony_cast_hint),
+            color = MaterialTheme.colorScheme.onSurfaceVariant,
+            style = MaterialTheme.typography.bodySmall,
+            textAlign = TextAlign.Center,
+        )
+    }
+}
+
+@Composable
+fun HarmonyConnectedDevices(tiles: List<TileViewModel>, modifier: Modifier = Modifier) {
+    val context = LocalContext.current
+    val defaults =
+        remember(context) {
+            mapOf(
+                BluetoothSpec to context.getString(R.string.quick_settings_bluetooth_label),
+                "hotspot" to context.getString(R.string.quick_settings_hotspot_label),
+            )
+        }
+    val candidates = remember(tiles) { tiles.filter { it.spec.spec in defaults } }
+    if (candidates.isEmpty()) return
+    ListenTo(candidates)
+    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = spacedBy(HarmonyGap)) {
+        candidates.forEach { tile ->
+            val (uiState, icon) = rememberTileState(tile)
+            val connected =
+                uiState.visualState == Tile.STATE_ACTIVE &&
+                    uiState.label.isNotBlank() &&
+                    uiState.label != defaults[tile.spec.spec]
+            if (connected) HarmonyDeviceCard(tile, uiState.label, uiState.secondaryLabel, icon)
+        }
+    }
+}
+
+@Composable
+private fun HarmonyDeviceCard(
+    tile: TileViewModel,
+    name: String,
+    status: String,
+    icon: Icon,
+) {
+    Row(
+        modifier =
+            Modifier.fillMaxWidth()
+                .harmonyCard()
+                .clickable { tile.mainClick(null) }
+                .padding(16.dp),
+        horizontalArrangement = spacedBy(16.dp),
+        verticalAlignment = Alignment.CenterVertically,
+    ) {
+        Box(
+            modifier =
+                Modifier.size(56.dp)
+                    .clip(RoundedCornerShape(14.dp))
+                    .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.06f)),
+            contentAlignment = Alignment.Center,
+        ) {
+            SmallTileContent(iconProvider = { icon }, color = HarmonyAccent, size = { 28.dp })
+        }
+        Column(Modifier.weight(1f)) {
+            Text(
+                text = name,
+                color = MaterialTheme.colorScheme.onSurface,
+                style = MaterialTheme.typography.titleSmall,
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis,
+            )
+            Text(
+                text =
+                    listOf(stringResource(R.string.harmony_connected), status)
+                        .filter { it.isNotBlank() }
+                        .joinToString(" | "),
+                color = MaterialTheme.colorScheme.onSurfaceVariant,
+                style = MaterialTheme.typography.bodySmall,
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis,
+            )
+        }
+    }
+}
diff --git a/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/EditMode.kt b/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/EditMode.kt
index a78656620693..401eda3d9d16 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/EditMode.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/EditMode.kt
@@ -141,6 +141,27 @@ private fun EditModeContent(viewModel: EditModeViewModel, modifier: Modifier = M
         remember(tiles, folderMemberSpecs) {
             tiles.filterNot { it.isCurrent && it.tileSpec.spec in folderMemberSpecs }
         }
+    if (panelStyle == QsPanelStyle.Harmony) {
+        val present = remember(tiles) { tiles.filter { it.isCurrent }.map { it.tileSpec.spec } }
+        val fixedSpecs =
+            remember(present) {
+                buildSet {
+                    listOf("wifi", "internet").firstOrNull { it in present }?.let(::add)
+                    if ("bt" in present) add("bt")
+                    if ("cast" in present) add("cast")
+                }
+            }
+        HarmonyEditMode(
+            tiles = tiles,
+            fixedSpecs = fixedSpecs,
+            onAddTile = viewModel::addTile,
+            onRemoveTile = viewModel::removeTile,
+            onSetTiles = viewModel::setTiles,
+            onStopEditing = viewModel::stopEditing,
+            modifier = modifier,
+        )
+        return
+    }
     Column(modifier) {
         gridLayout.EditTileGrid(
             if (panelElementsEditable) {
diff --git a/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/HarmonyEditMode.kt b/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/HarmonyEditMode.kt
new file mode 100644
index 000000000000..e9d5782aa46d
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/qs/panels/ui/compose/HarmonyEditMode.kt
@@ -0,0 +1,328 @@
+/*
+ * Copyright (C) 2026 The PenguinOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.systemui.qs.panels.ui.compose
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.gestures.detectDragGesturesAfterLongPress
+import androidx.compose.foundation.layout.Arrangement.spacedBy
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.BoxWithConstraints
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.offset
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.rememberScrollState
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.foundation.verticalScroll
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.automirrored.filled.ArrowBack
+import androidx.compose.material.icons.filled.Add
+import androidx.compose.material.icons.filled.Remove
+import androidx.compose.material3.Icon as MaterialIcon
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.key
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.draw.scale
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.text.style.TextOverflow
+import androidx.compose.ui.unit.IntOffset
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.zIndex
+import com.android.compose.theme.LocalAndroidColorScheme
+import com.android.systemui.qs.panels.ui.compose.infinitegrid.SmallTileContent
+import com.android.systemui.qs.panels.ui.viewmodel.EditTileViewModel
+import com.android.systemui.qs.pipeline.domain.interactor.CurrentTilesInteractor.Companion.POSITION_AT_END
+import com.android.systemui.qs.pipeline.shared.TileSpec
+import com.android.systemui.res.R
+import kotlin.math.roundToInt
+
+private const val Columns = 5
+private val CellHeight = 96.dp
+private val CircleSize = 52.dp
+private val CardCorner = 24.dp
+private val HarmonyAccent = Color(0xFF0A59F7)
+
+@Composable
+fun HarmonyEditMode(
+    tiles: List<EditTileViewModel>,
+    fixedSpecs: Set<String>,
+    onAddTile: (TileSpec, Int) -> Unit,
+    onRemoveTile: (TileSpec) -> Unit,
+    onSetTiles: (List<TileSpec>) -> Unit,
+    onStopEditing: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    val current = remember(tiles) { tiles.filter { it.isCurrent } }
+    val available = remember(tiles) { tiles.filterNot { it.isCurrent } }
+
+    Column(
+        modifier = modifier.fillMaxWidth().verticalScroll(rememberScrollState()),
+        verticalArrangement = spacedBy(16.dp),
+    ) {
+        Row(verticalAlignment = Alignment.CenterVertically) {
+            Box(
+                modifier = Modifier.size(48.dp).clip(CircleShape).clickable { onStopEditing() },
+                contentAlignment = Alignment.Center,
+            ) {
+                MaterialIcon(
+                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
+                    contentDescription = null,
+                    tint = MaterialTheme.colorScheme.onSurface,
+                )
+            }
+            Column(Modifier.padding(start = 8.dp)) {
+                Text(
+                    text = stringResource(R.string.harmony_edit_title),
+                    color = MaterialTheme.colorScheme.onSurface,
+                    style = MaterialTheme.typography.headlineSmall,
+                )
+                Text(
+                    text = stringResource(R.string.harmony_edit_subtitle),
+                    color = MaterialTheme.colorScheme.onSurfaceVariant,
+                    style = MaterialTheme.typography.bodySmall,
+                )
+            }
+        }
+
+        EditCard {
+            ReorderGrid(
+                tiles = current.filterNot { it.tileSpec.spec in fixedSpecs },
+                onRemoveTile = onRemoveTile,
+                onCommit = { order ->
+                    val moved = order.iterator()
+                    onSetTiles(
+                        current.map { tile ->
+                            if (tile.tileSpec.spec in fixedSpecs || !moved.hasNext()) tile.tileSpec
+                            else moved.next()
+                        }
+                    )
+                },
+            )
+        }
+
+        if (available.isNotEmpty()) {
+            Text(
+                text = stringResource(R.string.harmony_edit_more),
+                color = MaterialTheme.colorScheme.onSurfaceVariant,
+                style = MaterialTheme.typography.titleSmall,
+                modifier = Modifier.padding(start = 4.dp),
+            )
+            EditCard {
+                available.chunked(Columns).forEach { row ->
+                    Row(Modifier.fillMaxWidth()) {
+                        row.forEach { tile ->
+                            key(tile.tileSpec) {
+                                EditCell(
+                                    tile = tile,
+                                    badgeAdds = true,
+                                    onBadge = { onAddTile(tile.tileSpec, POSITION_AT_END) },
+                                    modifier = Modifier.weight(1f).height(CellHeight),
+                                )
+                            }
+                        }
+                        repeat(Columns - row.size) { Box(Modifier.weight(1f)) }
+                    }
+                }
+            }
+        }
+    }
+}
+
+@Composable
+private fun EditCard(content: @Composable () -> Unit) {
+    Column(
+        modifier =
+            Modifier.fillMaxWidth()
+                .clip(RoundedCornerShape(CardCorner))
+                .background(LocalAndroidColorScheme.current.surfaceEffect1.copy(alpha = 0.6f))
+                .padding(vertical = 12.dp, horizontal = 6.dp)
+    ) {
+        content()
+    }
+}
+
+@Composable
+private fun ReorderGrid(
+    tiles: List<EditTileViewModel>,
+    onRemoveTile: (TileSpec) -> Unit,
+    onCommit: (List<TileSpec>) -> Unit,
+) {
+    var order by remember { mutableStateOf(tiles) }
+    var dragged by remember { mutableIntStateOf(-1) }
+    var dragOffset by remember { mutableStateOf(Offset.Zero) }
+    LaunchedEffect(tiles) { if (dragged < 0) order = tiles }
+    val commit by rememberUpdatedState(onCommit)
+
+    BoxWithConstraints(Modifier.fillMaxWidth()) {
+        val density = LocalDensity.current
+        val cellWidthDp = maxWidth / Columns
+        val cellWidth = with(density) { cellWidthDp.toPx() }
+        val cellHeight = with(density) { CellHeight.toPx() }
+        val rows = (order.size + Columns - 1) / Columns
+        fun origin(index: Int) =
+            Offset((index % Columns) * cellWidth, (index / Columns) * cellHeight)
+
+        Box(Modifier.fillMaxWidth().height(CellHeight * rows.coerceAtLeast(1))) {
+            order.forEachIndexed { index, tile ->
+                key(tile.tileSpec) {
+                    val isDragged = index == dragged
+                    val position = origin(index) + if (isDragged) dragOffset else Offset.Zero
+                    EditCell(
+                        tile = tile,
+                        badgeAdds = false,
+                        onBadge =
+                            if (tile.isRemovable) {
+                                { onRemoveTile(tile.tileSpec) }
+                            } else {
+                                null
+                            },
+                        lifted = isDragged,
+                        modifier =
+                            Modifier.offset {
+                                    IntOffset(position.x.roundToInt(), position.y.roundToInt())
+                                }
+                                .size(cellWidthDp, CellHeight)
+                                .zIndex(if (isDragged) 1f else 0f)
+                                .pointerInput(tile.tileSpec) {
+                                    detectDragGesturesAfterLongPress(
+                                        onDragStart = {
+                                            dragged = order.indexOf(tile)
+                                            dragOffset = Offset.Zero
+                                        },
+                                        onDrag = { change, amount ->
+                                            change.consume()
+                                            val from = dragged
+                                            if (from < 0) return@detectDragGesturesAfterLongPress
+                                            dragOffset += amount
+                                            val centre =
+                                                origin(from) +
+                                                    dragOffset +
+                                                    Offset(cellWidth / 2, cellHeight / 2)
+                                            val column =
+                                                (centre.x / cellWidth).toInt().coerceIn(0, Columns - 1)
+                                            val row = (centre.y / cellHeight).toInt().coerceAtLeast(0)
+                                            val to =
+                                                (row * Columns + column).coerceIn(0, order.size - 1)
+                                            if (to != from) {
+                                                order =
+                                                    order.toMutableList().apply {
+                                                        add(to, removeAt(from))
+                                                    }
+                                                dragOffset -= origin(to) - origin(from)
+                                                dragged = to
+                                            }
+                                        },
+                                        onDragEnd = {
+                                            dragged = -1
+                                            dragOffset = Offset.Zero
+                                            commit(order.map { it.tileSpec })
+                                        },
+                                        onDragCancel = {
+                                            dragged = -1
+                                            dragOffset = Offset.Zero
+                                            commit(order.map { it.tileSpec })
+                                        },
+                                    )
+                                },
+                    )
+                }
+            }
+        }
+    }
+}
+
+@Composable
+private fun EditCell(
+    tile: EditTileViewModel,
+    badgeAdds: Boolean,
+    onBadge: (() -> Unit)?,
+    modifier: Modifier = Modifier,
+    lifted: Boolean = false,
+) {
+    Column(
+        modifier = modifier.padding(horizontal = 2.dp, vertical = 4.dp),
+        horizontalAlignment = Alignment.CenterHorizontally,
+        verticalArrangement = spacedBy(6.dp),
+    ) {
+        Box(Modifier.scale(if (lifted) 1.12f else 1f)) {
+            Box(
+                modifier =
+                    Modifier.size(CircleSize)
+                        .clip(CircleShape)
+                        .background(
+                            if (badgeAdds) MaterialTheme.colorScheme.onSurface.copy(alpha = 0.1f)
+                            else HarmonyAccent
+                        ),
+                contentAlignment = Alignment.Center,
+            ) {
+                SmallTileContent(
+                    iconProvider = { tile.icon },
+                    color = if (badgeAdds) MaterialTheme.colorScheme.onSurface else Color.White,
+                    size = { 24.dp },
+                )
+            }
+            if (onBadge != null) {
+                Box(
+                    modifier =
+                        Modifier.align(Alignment.TopEnd)
+                            .offset(x = 6.dp, y = (-6).dp)
+                            .size(22.dp)
+                            .clip(CircleShape)
+                            .background(MaterialTheme.colorScheme.surfaceContainerHighest)
+                            .clickable { onBadge() },
+                    contentAlignment = Alignment.Center,
+                ) {
+                    MaterialIcon(
+                        imageVector = if (badgeAdds) Icons.Filled.Add else Icons.Filled.Remove,
+                        contentDescription = null,
+                        tint = MaterialTheme.colorScheme.onSurface,
+                        modifier = Modifier.size(16.dp),
+                    )
+                }
+            }
+        }
+        Text(
+            text = tile.label.text,
+            color = MaterialTheme.colorScheme.onSurface,
+            style = MaterialTheme.typography.labelSmall,
+            maxLines = 2,
+            overflow = TextOverflow.Ellipsis,
+            textAlign = TextAlign.Center,
+        )
+    }
+}
diff --git a/packages/SystemUI/src/com/android/systemui/qs/shared/style/QsPanelStyle.kt b/packages/SystemUI/src/com/android/systemui/qs/shared/style/QsPanelStyle.kt
index c453972280e2..90f720d744ac 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/shared/style/QsPanelStyle.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/shared/style/QsPanelStyle.kt
@@ -21,7 +21,9 @@ enum class QsPanelStyle(val value: Int) {
 
     Penguin(1),
 
-    MyUi(2);
+    MyUi(2),
+
+    Harmony(3);
 
     companion object {
         const val SETTING_NAME = "qs_panel_style"
-- 
2.43.0

